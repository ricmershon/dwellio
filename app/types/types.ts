import { StructuredFormErrorMap } from "@/app/utils/build-form-error-map";

export interface Rates {
    nightly?: number;
    weekly?: number;
    monthly?: number
}

export type ActionState = {
    message?: string | null;
    status?: 'SUCCESS' | 'ERROR' | null,
    isFavorite?: boolean,
    isRead?: boolean,
    formData?: FormData,
    formErrorMap?: StructuredFormErrorMap
}

export interface PropertiesQuery {
    $or: [
        { name: RegExp },
        { description: RegExp },
        { amenities: RegExp },
        { type: RegExp },
        { 'location.street': RegExp },
        { 'location.city': RegExp },
        { 'location.state': RegExp },
        { 'location.zip': RegExp }
    ]
}

export interface PropertyImageData {
    secureUrl: string;
    publicId: string;
}

export type FormErrorsType = Record<string, string[]> | string[];

export const MAX_ITEMS_PER_PAGE = 6;