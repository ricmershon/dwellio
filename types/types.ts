import { StructuredFormErrorMap } from "@/utils/build-form-error-map";

export const ActionStatus = {
    INFO: 'info',
    SUCCESS: 'success',
    WARNING: 'warning',
    ERROR: 'error',
} as const;

type ActionStatus = (typeof ActionStatus)[keyof typeof ActionStatus];

export type ActionState = {
  message?: string | null;
  status?: ActionStatus | null;
  isFavorite?: boolean;
  isRead?: boolean;
  formData?: FormData;
  formErrorMap?: StructuredFormErrorMap;
};

export interface PropertiesQuery {
    $or: [
        { name: RegExp },
        { description: RegExp },
        { amenities: RegExp },
        { type: RegExp },
        { 'location.street': RegExp },
        { 'location.city': RegExp },
        { 'location.state': RegExp },
        { 'location.zip': RegExp }
    ]
}

export interface PropertyImageData {
    secureUrl: string;
    publicId: string;
    width: number;
    height: number;
}

export interface Rates {
    nightly?: number;
    weekly?: number;
    monthly?: number
}

export interface Amenity {
    id: string,
    value: string
}

export interface OptionType {
    value: string;
    label: string;
}

export type FormErrorsType = Record<string, string[]> | string[];

export const VIEWPORT_WIDTH_COOKIE_NAME = 'viewport_width';