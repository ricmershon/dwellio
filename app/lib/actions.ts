'use server';

import { revalidatePath } from "next/cache";
import { redirect } from "next/navigation";

import { Property } from "@/app/models/property-model";
import { PropertyInterfaceWithId } from "@/app/lib/definitions";
import connectDB from "@/app/config/database-config";
import { getSessionUser } from "@/app/utils/get-session-user";
import cloudinary, { uploadImages } from "@/app/lib/cloudinary";

export const addProperty = async (formData: FormData) => {
    await connectDB();

    const sessionUser = await getSessionUser();
    if (!sessionUser || !sessionUser.id) {
        throw new Error('User ID is required')
    }

    const propertyData = {
        owner: sessionUser.id,
        type: formData.get('type'),
        name: formData.get('name'),
        description: formData.get('description'),
        location: {
            street: formData.get('location.street'),
            city: formData.get('location.city'),
            state: formData.get('location.state'),
            zipcode: formData.get('location.zipcode'),
        },
        beds: formData.get('beds'),
        baths: formData.get('baths'),
        square_feet: formData.get('square_feet'),
        amenities: formData.getAll('amenities'),
        rates: {
            nightly: formData.get('rates.nightly'),
            weekly: formData.get('rates.weekly'),
            monthly: formData.get('rates.monthly')
        },
        seller_info: {
            name: formData.get('seller_info.name'),
            email: formData.get('seller_info.email'),
            phone: formData.get('seller_info.phone'),            
        }
    }

    const propertyImages = await uploadImages((formData.getAll('images') as Array<File>));
    Object.assign(propertyData, { images: propertyImages });

    const newProprety = new Property(propertyData);
    await newProprety.save();

    revalidatePath('/', 'layout');
    redirect(`/properties/${newProprety._id}`);
}

export const deleteProperty = async (propertyId: string) => {
    const sessionUser = await getSessionUser();
    if (!sessionUser || !sessionUser.id) {
        return { message: 'User ID is required.', status: 'ERROR' }
    }

    const property: PropertyInterfaceWithId | null = await Property.findById(propertyId);
    if (!property) {
        return { message: 'Property not found.', status: 'ERROR' }
    }

    // Verify ownwership
    if (property.owner.toString() !== sessionUser.id) {
        return { message: 'Not authorized to delete property.', status: 'ERROR' }
    }

    // Extract public ID from image URLs
    const imagePublicIds = property.images.map((imageUrl) => {
        const parts = imageUrl.split('/');
        return parts.at(-1)!.split('.').at(0);
    });

    // Delete images from Cloudinary
    if (imagePublicIds.length > 0) {
        for (const imagePublicId of imagePublicIds) {
            await cloudinary.uploader.destroy(`dwellio/${imagePublicId}`)
        }
    }

    await property.deleteOne();

    revalidatePath('/profile');
    return { message: 'Property successfully deleted.', status: 'SUCCESS' };
}