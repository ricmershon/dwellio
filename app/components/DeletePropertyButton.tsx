'use client';

import { toast } from "react-toastify";

import { deleteProperty } from "@/app/lib/actions";

const DeletePropertyButton = ({ propertyId }: { propertyId: string }) => {
    const deletePropertyById = deleteProperty.bind(null, propertyId);
    
    const deletePropertyAction = async () => {
        const result = await deletePropertyById();
        if (result.message) {
            if (result.status === 'SUCCESS') {
                toast.success(result.message, { position: 'bottom-right' });
            } else if (result.status === 'ERROR') {
                toast.error(result.message, { position: 'bottom-right' });
            }
        }
    }
    
    return (
        <form action={deletePropertyAction} className="inline-block">
            <button
                className="bg-red-500 text-white px-3 py-2 rounded-md hover:bg-red-600 cursor-pointer"
                type="submit"
            >
                Delete
            </button>
        </form>
    );
}
 
export default DeletePropertyButton;