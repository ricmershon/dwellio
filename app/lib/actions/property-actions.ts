'use server';

import { revalidatePath } from "next/cache";
import { redirect } from "next/navigation";
import { Types } from "mongoose";

import dbConnect from "@/app/config/database-config";
import type { ActionState, PropertyImageData } from "@/app/lib/definitions";
import { uploadImages, destroyImages } from "@/app/lib/cloudinary";
import { getSessionUser } from "@/app/utils/get-session-user";
import { toActionState } from "@/app/utils/to-action-state";
import { buildFormErrorMap } from "@/app/utils/build-form-error-map";
import { Property, PropertyDocument, User, UserDocument } from "@/app/models";
import { PropertyInput } from "@/app/schemas/property-schema";

// FIXME: fix difference between property id types
/**
 * Creates a new property.
 * 
 * @param {ActionState} _prevState - required by useActionState
 * @param {FormData} formData 
 * @returns Promise<ActionState> - ActionState may include form data in order to
 * repopulate the form if there's an error.
 */
export const createProperty = async (_prevState: ActionState, formData: FormData) => {
    const sessionUser = await getSessionUser();
    if (!sessionUser) {
        throw new Error('User ID is required.')
    }

    const rawImages = formData.getAll('images') as File[];
    const images = rawImages.filter(file => file.size > 0);

    const validationResults = PropertyInput.safeParse({
        type: formData.get('type'),
        name: formData.get('name'),
        description: formData.get('description'),
        location: {
            street: formData.get('location.street'),
            city: formData.get('location.city'),
            state: formData.get('location.state'),
            zipcode: formData.get('location.zipcode'),
        },
        beds: formData.get('beds'),
        baths: formData.get('baths'),
        squareFeet: formData.get('squareFeet'),
        amenities: formData.getAll('amenities'),
        rates: {
            nightly: formData.get('rates.nightly'),
            weekly: formData.get('rates.weekly'),
            monthly: formData.get('rates.monthly')
        },
        sellerInfo: {
            name: formData.get('sellerInfo.name'),
            email: formData.get('sellerInfo.email'),
            phone: formData.get('sellerInfo.phone'),            
        },
        imagesData: images
    });

    /**
     * Return immediately if form validation fails.
     */
    if (!validationResults.success) {
        const formErrorMap = buildFormErrorMap(validationResults.error.issues);
        return {
            formData: formData,
            formErrorMap: formErrorMap
        } as ActionState
    }

    /**
     * Create a property document that includes the data from `validationResults`
     * and removes the `imagesData` field and adds the `owner` field.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { imagesData, ...propertyDataWithoutImages } = validationResults.data;
    const propertyDocument = {
        ...propertyDataWithoutImages
    };
    Object.assign(propertyDocument, { owner: sessionUser.id });

    let newPropertyDocument: PropertyDocument;
    let newImagesData: PropertyImageData[] = [];

    try {
        await dbConnect();
        
        newImagesData = await uploadImages(rawImages);
        Object.assign(propertyDocument, { imagesData: newImagesData });
    
        newPropertyDocument = new Property(propertyDocument);
        await newPropertyDocument.save();
    } catch (error) {
        // Destroy the images just uploaded.
        if (newImagesData.length > 0) {
            destroyImages(newImagesData);
        }

        console.error(`>>> Database error adding a property: ${error}`);

        /**
         * Return form data so the form can be repopulated and the user does
         * not have to re-enter info.
         */
        return toActionState(
            `Failed to add a property: ${error}`,
            'ERROR',
            undefined,
            undefined,
            formData
        );
    }

    redirect(`/properties/${newPropertyDocument._id}`);
}

/**
 * Deletes a property.
 * 
 * @param {string} propertyId - id of the property to be deleted.
 * @returns Promise<ActionState>
 */
export const deleteProperty = async (propertyId: string) => {
    const sessionUser = await getSessionUser();
    if (!sessionUser || !sessionUser.id) {
        throw new Error('User ID is required.')
    }

    /**
     * Confirm properties existence and verify ownwerhip.
     */
    const property: PropertyDocument | null = await Property.findById(propertyId);
    if (!property) {
        return toActionState('Property not found.', 'ERROR');
    }

    if (property.owner.toString() !== sessionUser.id) {
        return toActionState('Not authorized to delete peoperty.', 'ERROR');
    }

    try {
        await dbConnect();

        destroyImages(property.imagesData!);
        await property.deleteOne();
    } catch (error) {
        console.error(`>>> Database error deleting a property: ${error}`);
        
        return toActionState(
            `Failed to add a property: ${error}`,
            'ERROR',
            undefined,
            undefined
        );
    }

    revalidatePath('/profile');
    return toActionState('Property successfully deleted.', 'SUCCESS');
}

/**
 * Updates a property.
 * 
 * @param {string} propertyId 
 * @param {ActionState} _prevState - required by useActionState
 * @param {FormData} formData 
 * @returns Promise<ActionState> - ActionState may include form data in order to
 * repopulate the form if there's an error.
 */
export const updateProperty = async (
    propertyId: string,
    _prevState: ActionState,
    formData: FormData
) => {
    const sessionUser = await getSessionUser();
    if (!sessionUser || !sessionUser.id) {
        throw new Error('User ID is required.')
    }

    /**
     * Confirm property's existence and verify ownership.
     */
    const property: PropertyDocument | null = await Property.findById(propertyId);
    if (!property) {
        return toActionState('Property not found.', 'ERROR');
    }

    if (property.owner.toString() !== sessionUser.id) {
        return toActionState('Not authorized to update property.', 'ERROR');
    }

    /**
     * Remove `imagesData` from the data validation, extract property data and
     * update property.
     */
    const UpdateProperty = PropertyInput.omit({ imagesData: true });
    const validationResults = UpdateProperty.safeParse({
        type: formData.get('type'),
        name: formData.get('name'),
        description: formData.get('description'),
        location: {
            street: formData.get('location.street'),
            city: formData.get('location.city'),
            state: formData.get('location.state'),
            zipcode: formData.get('location.zipcode'),
        },
        beds: formData.get('beds'),
        baths: formData.get('baths'),
        squareFeet: formData.get('squareFeet'),
        amenities: formData.getAll('amenities'),
        rates: {
            nightly: formData.get('rates.nightly'),
            weekly: formData.get('rates.weekly'),
            monthly: formData.get('rates.monthly')
        },
        sellerInfo: {
            name: formData.get('sellerInfo.name'),
            email: formData.get('sellerInfo.email'),
            phone: formData.get('sellerInfo.phone'),            
        }
    });

    /**
     * Return immediately if form validation fails.
     */
    if (!validationResults.success) {
        const formErrorMap = buildFormErrorMap(validationResults.error.issues);
        return {
            formData: formData,
            formErrorMap: formErrorMap
        } as ActionState
    }

    try {
        await dbConnect();

        await Property.findByIdAndUpdate(propertyId, validationResults.data);
    } catch (error) {
        console.error(`>>> Database error updating a property: ${error}`);

        /**
         * Return form data so the form can be repopulated and the user does
         * not have to re-enter info.
         */
        return toActionState(
            `Failed to add a property: ${error}`,
            'ERROR',
            undefined,
            undefined,
            formData
        );
    }

    revalidatePath('/', 'layout');
    redirect(`/properties/${propertyId}`);
}

/**
 * Sets favorite on a property.
 * 
 * @param {string} propertyId 
 * @returns Promise<ActionState>
 */
export const favoriteProperty = async (propertyId: string) => {
    const propertyObjectId = new Types.ObjectId(propertyId);
    
    await dbConnect();

    const sessionUser = await getSessionUser();
    if (!sessionUser || !sessionUser.id) {
        throw new Error('User ID is required.')
    }

    let user: UserDocument | null;
    let actionState: ActionState;

    try {
        user = await User.findById(sessionUser.id);
        if (!user) {
            return toActionState('User not found.', 'ERROR');
        } 
    } catch (error) {
        console.error(`>>> Database error finding user: ${error}`);
        return toActionState(
            `Error finding user: ${error}`,
            'ERROR',
        );
    }

    const isFavorite = user.favorites.includes(propertyObjectId);

    try {
        let updatedUser;

        if (isFavorite) {
            updatedUser = await User.updateOne(
                { _id: user._id },
                { $pull: { favorites: propertyId } }
            );
            actionState = {
                message: 'Removed from favorites',
                status: 'SUCCESS',
                isFavorite: false
            }
        } else {
            updatedUser = await User.updateOne(
                { _id: user._id },
                { $push: { favorites: propertyId } }
            );
            actionState = {
                message: 'Added to favorites',
                status: 'SUCCESS',
                isFavorite: true
            }
        }
        if (updatedUser.modifiedCount !== 1) {
            console.error(`>>> Database error favoriting property`);
            return toActionState('Error favoriting property', 'ERROR');
        }
    } catch (error) {
        console.error(`>>> Database error favoriting property: ${error}`)
        return toActionState(`Error favoriting property`, 'ERROR');
    }

    revalidatePath('/properties/favorites', 'page');
    return actionState;
}

/**
 * Gets favorite status for a property.
 * 
 * @param {string} propertyId - property on which favorite is being set or unset.
 * @returns Promise<ActionState>
 */
export const getFavoriteStatus = async (propertyId: string) => {
    const propertyObjectId = new Types.ObjectId(propertyId);
    
    const sessionUser = await getSessionUser();
    if (!sessionUser || !sessionUser.id) {
        throw new Error('User ID is required.')
    }
    
    let user: UserDocument | null;
    
    try {
        await dbConnect();
        user = await User.findById(sessionUser.id);
        if (!user) {
            return toActionState('User not found.', 'ERROR');
        } 
    } catch (error) {
        console.error(`>>> Database error finding user: ${error}`)
        return toActionState(
            `Error finding user: ${error}`,
            'ERROR',
        );
    }

    const isFavorite = user.favorites.includes(propertyObjectId);

    return toActionState('Successfully fetched favorites status', 'SUCCESS', isFavorite);
}