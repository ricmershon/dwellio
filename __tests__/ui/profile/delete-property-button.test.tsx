import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { toast } from 'react-toastify';
import DeletePropertyButton from '@/ui/profile/delete-property-button';
import { deleteProperty } from '@/lib/actions/property-actions';
import { ActionStatus } from '@/types/types';

// Mock dependencies
jest.mock('react-toastify', () => ({
    toast: {
        success: jest.fn(),
        error: jest.fn(),
        info: jest.fn(),
        warning: jest.fn()
    }
}));

jest.mock('@/lib/actions/property-actions', () => ({
    deleteProperty: jest.fn()
}));

const mockDeleteProperty = deleteProperty as jest.MockedFunction<typeof deleteProperty>;
const mockToast = toast as jest.Mocked<typeof toast>;

describe('DeletePropertyButton Component', () => {
    const mockPropertyId = 'property123';

    beforeEach(() => {
        jest.clearAllMocks();
    });

    describe('Component Initialization', () => {
        it('should accept propertyId prop correctly', () => {
            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const button = screen.getByRole('button', { name: 'Delete' });
            expect(button).toBeInTheDocument();
        });

        it('should render form with inline-block styling', () => {
            const { container } = render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const form = container.querySelector('form');
            expect(form).toHaveClass('inline-block');
        });

        it('should render button with correct type and styling', () => {
            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const button = screen.getByRole('button', { name: 'Delete' });
            expect(button).toHaveAttribute('type', 'submit');
            expect(button).toHaveClass('btn', 'btn-danger');
        });

        it('should create bound action function properly', () => {
            // This test verifies the component renders without errors, 
            // indicating the bind operation worked
            expect(() => {
                render(<DeletePropertyButton propertyId={mockPropertyId} />);
            }).not.toThrow();
        });
    });

    describe('Form Submission', () => {
        it('should call deleteProperty action on form submit', async () => {
            const user = userEvent.setup();
            mockDeleteProperty.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Property deleted successfully'
            });

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const button = screen.getByRole('button', { name: 'Delete' });
            await user.click(button);

            expect(mockDeleteProperty).toHaveBeenCalledTimes(1);
        });

        it('should handle form submission events properly', async () => {
            const user = userEvent.setup();
            mockDeleteProperty.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Property deleted successfully'
            });

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const form = screen.getByRole('button', { name: 'Delete' }).closest('form');
            expect(form).toBeInTheDocument();

            await user.click(screen.getByRole('button', { name: 'Delete' }));

            expect(mockDeleteProperty).toHaveBeenCalled();
        });

        it('should execute bound action with correct propertyId', async () => {
            const user = userEvent.setup();
            
            // Mock the deleteProperty to capture what it was called with
            const mockBoundAction = jest.fn().mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Property deleted'
            });
            
            // Since we're testing the bound function, we need to mock the implementation
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            mockDeleteProperty.mockImplementation(() => Promise.resolve(mockBoundAction()) as Promise<any>);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            // The deleteProperty should have been used in the bind operation
            expect(mockDeleteProperty).toHaveBeenCalled();
        });

        it('should handle multiple rapid clicks gracefully', async () => {
            const user = userEvent.setup();
            mockDeleteProperty.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Property deleted'
            });

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const button = screen.getByRole('button', { name: 'Delete' });
            
            // Click multiple times rapidly
            await user.click(button);
            await user.click(button);
            await user.click(button);

            // Should handle multiple clicks without errors
            expect(mockDeleteProperty).toHaveBeenCalled();
        });
    });

    describe('Server Action Integration', () => {
        it('should handle successful deletion response', async () => {
            const user = userEvent.setup();
            const successResult = {
                status: ActionStatus.SUCCESS,
                message: 'Property deleted successfully'
            };
            
            mockDeleteProperty.mockResolvedValue(successResult);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            await waitFor(() => {
                expect(mockToast.success).toHaveBeenCalledWith('Property deleted successfully');
            });
        });

        it('should handle error deletion response', async () => {
            const user = userEvent.setup();
            const errorResult = {
                status: ActionStatus.ERROR,
                message: 'Failed to delete property'
            };
            
            mockDeleteProperty.mockResolvedValue(errorResult);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            await waitFor(() => {
                expect(mockToast.error).toHaveBeenCalledWith('Failed to delete property');
            });
        });

        it('should process ActionState response format correctly', async () => {
            const user = userEvent.setup();
            const result = {
                status: ActionStatus.SUCCESS,
                message: 'Operation completed',
                formData: new FormData()
            };
            
            mockDeleteProperty.mockResolvedValue(result);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            await waitFor(() => {
                expect(mockToast[result.status]).toHaveBeenCalledWith(result.message);
            });
        });

        it('should handle action responses without status', async () => {
            const user = userEvent.setup();
            const resultWithoutStatus = {
                message: 'Some message'
            };
            
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            mockDeleteProperty.mockResolvedValue(resultWithoutStatus as any);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            // Should not crash when status is missing
            await waitFor(() => {
                expect(mockDeleteProperty).toHaveBeenCalled();
            });
        });
    });

    describe('Toast Notifications', () => {
        it('should use toast[result.status] dynamic pattern', async () => {
            const user = userEvent.setup();
            const result = {
                status: ActionStatus.SUCCESS,
                message: 'Dynamic toast test'
            };
            
            mockDeleteProperty.mockResolvedValue(result);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            await waitFor(() => {
                expect(mockToast[ActionStatus.SUCCESS]).toHaveBeenCalledWith('Dynamic toast test');
            });
        });

        it('should display result.message as toast content', async () => {
            const user = userEvent.setup();
            const customMessage = 'Custom deletion message';
            const result = {
                status: ActionStatus.SUCCESS,
                message: customMessage
            };
            
            mockDeleteProperty.mockResolvedValue(result);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            await waitFor(() => {
                expect(mockToast.success).toHaveBeenCalledWith(customMessage);
            });
        });

        it('should handle different toast types based on status', async () => {
            const user = userEvent.setup();
            
            const { rerender } = render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            // Test success toast
            mockDeleteProperty.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Success message'
            });
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));
            
            await waitFor(() => {
                expect(mockToast.success).toHaveBeenCalledWith('Success message');
            });

            // Reset mocks for next test
            jest.clearAllMocks();
            
            // Test error toast
            mockDeleteProperty.mockResolvedValue({
                status: ActionStatus.ERROR,
                message: 'Error message'
            });
            
            rerender(<DeletePropertyButton propertyId="different-property" />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));
            
            await waitFor(() => {
                expect(mockToast.error).toHaveBeenCalledWith('Error message');
            });
        });

        it('should handle missing message gracefully', async () => {
            const user = userEvent.setup();
            const resultWithoutMessage = {
                status: ActionStatus.SUCCESS
            };
            
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            mockDeleteProperty.mockResolvedValue(resultWithoutMessage as any);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            await waitFor(() => {
                expect(mockToast.success).not.toHaveBeenCalled();
            });
        });
    });

    describe('Action Results Handling', () => {
        it('should validate result object structure', async () => {
            const user = userEvent.setup();
            const validResult = {
                status: ActionStatus.SUCCESS,
                message: 'Valid result'
            };
            
            mockDeleteProperty.mockResolvedValue(validResult);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            await waitFor(() => {
                expect(mockToast.success).toHaveBeenCalledWith('Valid result');
            });
        });

        it('should handle undefined result gracefully', async () => {
            const user = userEvent.setup();
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            mockDeleteProperty.mockResolvedValue(undefined as any);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);

            // Should not crash with undefined result
            expect(() => {
                user.click(screen.getByRole('button', { name: 'Delete' }));
            }).not.toThrow();
        });

        it('should handle null result gracefully', async () => {
            const user = userEvent.setup();
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            mockDeleteProperty.mockResolvedValue(null as any);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);

            // Should not crash with null result
            expect(() => {
                user.click(screen.getByRole('button', { name: 'Delete' }));
            }).not.toThrow();
        });

        it('should handle result with extra properties', async () => {
            const user = userEvent.setup();
            const resultWithExtras = {
                status: ActionStatus.SUCCESS,
                message: 'Success with extras',
                extraProperty: 'should be ignored',
                anotherExtra: 123
            };
            
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            mockDeleteProperty.mockResolvedValue(resultWithExtras as any);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            await waitFor(() => {
                expect(mockToast.success).toHaveBeenCalledWith('Success with extras');
            });
        });
    });

    describe('Button Rendering', () => {
        it('should render button with "Delete" text', () => {
            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            expect(screen.getByRole('button', { name: 'Delete' })).toBeInTheDocument();
        });

        it('should apply danger button styling', () => {
            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const button = screen.getByRole('button', { name: 'Delete' });
            expect(button).toHaveClass('btn', 'btn-danger');
        });

        it('should set button type to submit', () => {
            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const button = screen.getByRole('button', { name: 'Delete' });
            expect(button).toHaveAttribute('type', 'submit');
        });

        it('should maintain consistent button appearance', () => {
            const { container } = render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const button = container.querySelector('button');
            expect(button).toHaveClass('btn', 'btn-danger');
            expect(button).toHaveAttribute('type', 'submit');
            expect(button?.textContent?.trim()).toBe('Delete');
        });
    });

    describe('Error Scenarios', () => {
        it('should handle server action failures gracefully', async () => {
            const user = userEvent.setup();
            mockDeleteProperty.mockRejectedValue(new Error('Server error'));

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            // Should not crash on server errors
            expect(() => {
                user.click(screen.getByRole('button', { name: 'Delete' }));
            }).not.toThrow();
        });

        it('should handle network connectivity issues', async () => {
            mockDeleteProperty.mockRejectedValue(new Error('Network error'));

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            // Component should remain functional
            expect(screen.getByRole('button', { name: 'Delete' })).toBeInTheDocument();
        });

        it('should handle authorization errors', async () => {
            const user = userEvent.setup();
            const authError = {
                status: ActionStatus.ERROR,
                message: 'Unauthorized: You do not own this property'
            };
            
            mockDeleteProperty.mockResolvedValue(authError);

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            await user.click(screen.getByRole('button', { name: 'Delete' }));

            await waitFor(() => {
                expect(mockToast.error).toHaveBeenCalledWith('Unauthorized: You do not own this property');
            });
        });

        it('should maintain UI state on errors', async () => {
            mockDeleteProperty.mockRejectedValue(new Error('Some error'));

            render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            const button = screen.getByRole('button', { name: 'Delete' });
            
            // Button should remain clickable after error
            expect(button).toBeEnabled();
            expect(button).toHaveClass('btn', 'btn-danger');
        });
    });

    describe('Multiple Instances', () => {
        it('should handle multiple DeletePropertyButton instances', () => {
            render(
                <div>
                    <DeletePropertyButton propertyId="property1" />
                    <DeletePropertyButton propertyId="property2" />
                    <DeletePropertyButton propertyId="property3" />
                </div>
            );
            
            const buttons = screen.getAllByRole('button', { name: 'Delete' });
            expect(buttons).toHaveLength(3);
        });

        it('should maintain independent state for multiple instances', async () => {
            const user = userEvent.setup();
            mockDeleteProperty.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Deleted'
            });

            render(
                <div>
                    <DeletePropertyButton propertyId="property1" />
                    <DeletePropertyButton propertyId="property2" />
                </div>
            );
            
            const buttons = screen.getAllByRole('button', { name: 'Delete' });
            
            // Click first button
            await user.click(buttons[0]);
            
            // Both buttons should remain functional
            expect(buttons[0]).toBeEnabled();
            expect(buttons[1]).toBeEnabled();
        });
    });

    describe('Performance and Memory', () => {
        it('should not cause memory leaks with multiple renders', () => {
            const { rerender, unmount } = render(<DeletePropertyButton propertyId={mockPropertyId} />);
            
            // Re-render multiple times
            for (let i = 0; i < 10; i++) {
                rerender(<DeletePropertyButton propertyId={`property${i}`} />);
            }
            
            // Should unmount cleanly
            expect(() => unmount()).not.toThrow();
        });

        it('should handle rapid re-renders without issues', () => {
            const { rerender } = render(<DeletePropertyButton propertyId="initial" />);
            
            // Rapid re-renders
            rerender(<DeletePropertyButton propertyId="updated1" />);
            rerender(<DeletePropertyButton propertyId="updated2" />);
            rerender(<DeletePropertyButton propertyId="updated3" />);
            
            expect(screen.getByRole('button', { name: 'Delete' })).toBeInTheDocument();
        });
    });

    describe('Snapshots', () => {
        it('should match snapshot with default button state', () => {
            const { container } = render(<DeletePropertyButton propertyId={mockPropertyId} />);
            expect(container.firstChild).toMatchSnapshot();
        });

        it('should match snapshot with different propertyId', () => {
            const { container } = render(<DeletePropertyButton propertyId="different-property-id" />);
            expect(container.firstChild).toMatchSnapshot();
        });

        it('should match snapshot showing form layout structure', () => {
            const { container } = render(<DeletePropertyButton propertyId={mockPropertyId} />);
            expect(container.firstChild).toMatchSnapshot();
        });
    });
});