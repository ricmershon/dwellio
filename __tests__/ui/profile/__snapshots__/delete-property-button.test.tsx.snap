// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`DeletePropertyButton Component Snapshots should match snapshot showing form layout structure 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="inline-block"
>
  <button
    class="btn btn-danger"
    type="submit"
  >
    Delete
  </button>
</form>
`;

exports[`DeletePropertyButton Component Snapshots should match snapshot with default button state 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="inline-block"
>
  <button
    class="btn btn-danger"
    type="submit"
  >
    Delete
  </button>
</form>
`;

exports[`DeletePropertyButton Component Snapshots should match snapshot with different propertyId 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="inline-block"
>
  <button
    class="btn btn-danger"
    type="submit"
  >
    Delete
  </button>
</form>
`;
